docs/design/2022-06-30-partition-table-batchwrite-suppport.md

```diff
@@ -0,0 +1,121 @@
+
+### The feature-level implementation of partitioning in TiDB
+
+Currently, TiDB supports Range partitioning, List partitioning, List COLUMNS partitioning, and Hash
+partitioning.
+For a table partitioned by [RANGE COLUMNS](https://dev.mysql.com/doc/refman/5.7/en/partitioning-columns-range.html),
+currently TiDB only supports using a single partitioning column.
+You can read the [documentation](https://docs.pingcap.com/tidb/dev/partitioned-table#list-partitioning) for more
+details.
+
+### The code-level implementation of partitioning in TiDB
+
+If a table is partitioned,
+the [`TableInfo`](https://github.com/pingcap/tidb/blob/v6.1.0/parser/model/model.go#L370-L447) holds
+the id of the single table by the SQL layer, which can be called logical tableId,
+while the n(n is the number of partitions)
+[`PartitionDefinition`](https://github.com/pingcap/tidb/blob/v6.1.0/parser/model/model.go#L1076-L1084)
+each holds the id of the table stored in TiKV, which can be called physical tableId.
+
+It's typical that the logical tableId should be used to allocate id or gain meta information, while
+the physical tableId should be used to generate encoded key or any scans which we operate on the
+physical table.
+
+## Goals
+
+Since TiSpark supports INSERT, REPLACE and DELETE semantics with DataSource API, it's natural to
+support the partitioning feature in these semantics.
+List partitioning, List COLUMNS partitioning are generally available in V6.1 recently and also more
+complicated. Considering the workload of this feature, we have decide to support them in the future.
+In this feature, we only focus on Range partitioning and Hash partitioning.
+
+## API changes
+
+There is no API change in this feature. Users use the DataSource API to execute INSERT, REPLACE and
+DELETE as normal.
+
+## Detailed design
+
+Since TiSpark supports INSERT, REPLACE and DELETE semantics with the physical table, the
+responsibility is simplified to transform the `TableInfo` of the partition table to the physical
+table so that we can reuse the existing API rather than bring more complexity into the code.
+
+### Generate the expression for partitioning
+
+The original type of expression is `String`, we need to use `TiParser` to parse the expression in
+order to evaluate the partitioning expression for the latter steps.
+
+- For the `Range` and `Range Column`  partitioning, it also needs to extract the less than parts
+  from 'partition p0 less than xx ... partition p1 less than ...'
+
+### Locate the partition for each row
+
+In this step, we need to get the physical tableId for each row.
+
+For the `Range` partitioning, the expression can be a column or indeed a function.
+
+- If the expression is a column, we can use the column's value to compare with the less than parts
+  to locate the partition.
+- If the expression is a function, we can evaluate the function's value and then compare with the
+  less than parts to locate the partition
```

Comment:
is there any limit for the function? we'd better clear it if it has some limits of function or we can list the supported functions

Reply:
I look up the user guide of TiDB and MySQL and there is nowhere mentioned the limit.
So it's supposed to no limit, any function TiDB supported should work fine

Reply:
The limit is the result of function must be integer

Reply:
But for this pr, only YEAR() function is supported